### PR TITLE
Reset password

### PR DESCRIPTION
**Pull request by Drvanon**
This should resolve issue #50.

### CODE REVIEW

File: web/gridt/schemas.py [hunk 1]
```diff
@@ -19,3 +19,17 @@ class MovementSchema(Schema):
...
+    new_password = fields.Str(required=True)
+
+
+class RequestPasswordResetSchema(Schema):
+    email = fields.Email(required=True)
+
+class ResetPasswordSchema(Schema):
+    token = fields.Str(required=True)
+    password = fields.Str(required=True)
+    password2 = fields.Str(required=True, validate=Equal(password))
```
Comment (Drvanon) [inline web/gridt/schemas.py:35]: This should be done only on the front end.

File: web/gridt/resources/user.py [hunk 2]
```diff
@@ -24,3 +31,56 @@ def put(self):
...
+    def post(self):
+
+        try:
+            res = self.schema.load(request.get_json())
+        except ValidationError as error:
+            field = list(error.messages.keys())[0]
+            return ({"message": f"{field}: {error.messages[field][0]}"}, 400)
+        
+        if not res["password"] == res["password2"]:
+            return ({"message": "Passwords do not match."}, 400)
```
Comment (Drvanon) [inline web/gridt/resources/user.py:83]: This would be double the work that the Schema does.

File: web/gridt/resources/user.py [hunk 3]
```diff
@@ -24,3 +31,56 @@ def put(self):
...
+
+        if
```
Comment (Drvanon) [inline web/gridt/resources/user.py:85]: You are not done here.

File: web/gridt/resources/user.py [hunk 4]
```diff
@@ -24,3 +31,56 @@ def put(self):
...
+
+        try:
+            res = self.schema.load(request.get_json())
+        except ValidationError as error:
+            field = list(error.messages.keys())[0]
+            return ({"message": f"{field}: {error.messages[field][0]}"}, 400)
+
+        if not current_app.config["EMAIL_API_KEY"]:
+            return ({"message": "Could not send e-mail: API key not available."}, 500)
+
```
Comment (Drvanon) [inline web/gridt/resources/user.py:69]: Logic is missing here still.

File: web/gridt/resources/user.py (hunk 3 above)
Comment (Drvanon) [reply web/gridt/resources/user.py:85]: Side note: normally it is better to not commit code that is not syntactically correct.

File: web/gridt/tests/integration/resources/test_user.py [hunk 5]
```diff
@@ -1,18 +1,25 @@
...
+import datetime
+
 from gridt.tests.base_test import BaseTest
 from gridt.models.user import User
-from gridt.resources.user import BioResource
+from gridt.resources.user import (
+    BioResource,
+    ChangePasswordResource,
+    RequestPasswordResetResource
+)
```
Comment (Drvanon) [inline web/gridt/tests/integration/resources/test_user.py:14]: Are you using these somewhere?

File: web/gridt/tests/integration/resources/test_user.py [hunk 6]
```diff
@@ -23,9 +30,214 @@ def test_update_bio(self):
...
+                    "old_password": "gibberish",
+                    "new_password": "somethingyoullneverguess",
+                },
+            )
+
+            self.assertIn("message", resp.get_json())
+            self.assertEqual(resp.status_code, 400)
+
+    def test_change_password_with_correct_password(self):
+        # json should contain old_password (self.users[0]["password"]) and new password
```
Comment (Drvanon) [inline web/gridt/tests/integration/resources/test_user.py:78]: You can give this text as a second input in the function. Then this text will show up when the test fails.
https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertEqual

Comment (Drvanon) [reply web/gridt/tests/integration/resources/test_user.py:78]: Maybe this is the wrong position to write that text.

File: web/gridt/tests/integration/resources/test_user.py [hunk 7]
```diff
@@ -23,9 +30,214 @@ def test_update_bio(self):
...
+                print(token)
+
+            # User uses expired token
+            with freeze_time("2020-04-19 22:10:00"):
+                resp = self.client.post(
+                    "/reset_password",
+                    json={
+                        "token": token,
+                        "password": "testpass",
+                        "password": "testpass"
```
Comment (Drvanon) [inline web/gridt/tests/integration/resources/test_user.py:202]: This does not make sense.

File: web/gridt/tests/integration/resources/test_user.py (hunk 6 above)
Comment (PJvanVliet) [reply web/gridt/tests/integration/resources/test_user.py]: Which lines does this refer to?

File: web/gridt/tests/integration/resources/test_user.py (hunk 7 above)
Comment (PJvanVliet) [reply web/gridt/tests/integration/resources/test_user.py:202]: That was a typo, will be solved since json now only needs "token" and "password" entries, no "password2".

File: web/gridt/tests/integration/resources/test_user.py (hunk 6 above)
Comment (Drvanon) [reply web/gridt/tests/integration/resources/test_user.py]: Line 78

File: web/gridt/resources/user.py [hunk 8]
```diff
@@ -24,3 +31,56 @@ def put(self):
...
+    schema = ChangePasswordSchema()
+
+    @jwt_required()
+    def post(self):
+
+        try:
+            res = self.schema.load(request.get_json())
+        except ValidationError as error:
+            field = list(error.messages.keys())[0]
+            return ({"message": f"{field}: {error.messages[field][0]}"}, 400)
```
[the remark this answers is not in the document]
Comment (Drvanon) [reply web/gridt/resources/user.py]: We could create our own "base resource" class which is then inherited by the actual resources.

File: web/docs/_static/openapi.yaml [hunk 9]
```diff
@@ -136,6 +136,73 @@ paths:
...
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/message'
+        '500':
+          description: Could not send e-mail; API key not available.
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/message'
```
Comment (Drvanon) [inline web/docs/_static/openapi.yaml:172]: Technically not wrong to put this here, although it is a little redundant.

File: web/gridt/resources/user.py [hunk 10]
```diff
@@ -45,4 +53,71 @@ def post(self):
...
+
+        try:
+            token_decoded = jwt.decode(res["token"], secret_key, algorithms=["HS256"])
+        except jwt.ExpiredSignatureError:
+            msg = "Signature has expired."
+            return ({"message": msg}, 400)
+        except jwt.InvalidTokenError:
+            msg = "Invalid token."
+            return ({"message": msg}, 400)
```
Comment (Drvanon) [inline web/gridt/resources/user.py:109]: It is actually much more logical to put this in the schema, as it has to do with the direct validation of the input. It could be dealt with there.

File: web/gridt/resources/user.py [hunk 11]
```diff
@@ -45,4 +53,71 @@ def post(self):
...
+        
+        id = token_decoded["reset_password"]
```
Comment (Drvanon) [inline web/gridt/resources/user.py:111]: Why is this called `reset_password` and not just `id`?

File: web/gridt/resources/user.py [hunk 12]
```diff
@@ -45,4 +53,71 @@ def post(self):
...
+        password = res["password"]
+        user = User.query.get(id)
+        user.hash_password(password)
+
+        link = "https://app.gridt.org/request_password_reset"
```
Comment (Drvanon) [inline web/gridt/resources/user.py:116]: This link doesn't work right now. Have you communicated this with the client team?

File: web/gridt/resources/user.py [hunk 13]
```diff
@@ -45,4 +53,71 @@ def post(self):
...
+        subj = "Your password has been changed"
+        body = f"Dear user, your password has been changed. Didn't do this? \
```
Comment (Drvanon) [inline web/gridt/resources/user.py:118]: You can actually write HTML in SendGrid emails, which allows you to make actual links and prettify things.

File: web/gridt/schemas.py [hunk 14]
```diff
@@ -24,3 +24,11 @@ class MovementSchema(Schema):
...
     old_password = fields.Str(required=True)
     new_password = fields.Str(required=True)
+
+
+class RequestPasswordResetSchema(Schema):
+    email = fields.Email(required=True)
+
+class ResetPasswordSchema(Schema):
+    token = fields.Str(required=True)
+    password = fields.Str(required=True)
```
Comment (Drvanon) [inline web/gridt/schemas.py:34]: Maybe move the JWT logic here.

File: web/gridt/tests/integration/resources/test_user.py [hunk 15]
```diff
@@ -80,7 +86,133 @@ def test_change_password_with_correct_password(self):
...
+                Secure your account by going to {link}."
 
             self.assertIn("message", resp.get_json())
             self.assertEqual(resp.status_code, 200)
-            self.assertTrue(user.verify_password("somethingyoullneverguess"))
\ No newline at end of file
+            self.assertTrue(user.verify_password("somethingyoullneverguess"))
+            func.assert_called_with(user.email, subj, body)
+
+
```
Comment (Drvanon) [inline web/gridt/tests/integration/resources/test_user.py:100]: Unnecessary whitespace.

Comment (Drvanon) [reply web/gridt/tests/integration/resources/test_user.py:100]: Consider running black.

File: web/gridt/tests/unit/models/test_user.py [hunk 16]
```diff
@@ -31,3 +35,20 @@ def test_hash(self):
     def test_avatar(self):
         user = User("username", "test@test.com", "test")
         self.assertEqual(user.get_email_hash(), "b642b4217b34b1e8d3bd915fc65c4452")
+
+    def test_get_password_reset_token(self):
+        with self.app_context():
+            user = User("username", "test@test.com", "password")
+            current_app.config["SECRET_KEY"] = "test"
```
Comment (Drvanon) [inline web/gridt/tests/unit/models/test_user.py:42]: Why did you set this here?

**Review by Drvanon — COMMENTED** [review]
Many things are looking good, there is just a few style things that I would like to have changed. 

All this work makes me consider that it might be good to work with blueprints, as this all kind of belongs to one subgroup. I don't know if we could make that work with flask-restful, but we should do some testing with it. Another thing that we might consider is putting all of these endpoints under one main endpoint, like /user/change-email, /user/change-password/request, /user/change-password/confirm, /user/reset-password. That could be neater, possibly.

File: web/gridt/resources/user.py (hunk 11 above)
Comment (PJvanVliet) [reply web/gridt/resources/user.py:111]: Renamed `user_id` instead, to be more specific.

File: web/gridt/resources/user.py (hunk 12 above)
Comment (PJvanVliet) [reply web/gridt/resources/user.py:116]: I'm on it, quite the oversight on my part.

File: web/gridt/tests/unit/models/test_user.py (hunk 16 above)
Comment (PJvanVliet) [reply web/gridt/tests/unit/models/test_user.py:42]: Not strictly necessary, if we assume that the secret key is set correctly.

File: web/gridt/schemas.py (hunk 14 above)
Comment (PJvanVliet) [reply web/gridt/schemas.py]: As of yet I have no idea how to implement this with marshmallow.

**Comment by Drvanon** [comment]
Could you fix the merge conflict?

**Review by Drvanon — COMMENTED** [review]
Looking good!

File: web/gridt/resources/user.py [hunk 17]
```diff
@@ -42,7 +53,54 @@ def post(self):
...
+        except ValidationError as error:
+            field = list(error.messages.keys())[0]
+            return ({"message": f"{field}: {error.messages[field][0]}"}, 400)
+
+        token_decoded = jwt.decode(res["token"], secret_key, algorithms=["HS256"])
+        user_id = token_decoded["user_id"]
+        password = res["password"]
+
+        user = User.find_by_id(user_id)
+        user.hash_password(password)
```
Comment (Drvanon) [inline web/gridt/resources/user.py:103]: This function should be renamed to hash_and_store_password or something like that, as it is not entirely clear from just reading what you are doing here.

File: web/util/send_email.py [hunk 18]
```diff
@@ -0,0 +1,20 @@
...
+    msg.template_id = template_id
+    msg.template_data = template_data
+
+    try:
+        sg = SendGridAPIClient(current_app.config["EMAIL_API_KEY"])
+        resp = sg.send(msg)
+        return resp.status_code, resp.body, resp.headers
+    except Exception as e:
+        if e.message:
+            return e.message
```
Comment (Drvanon) [inline web/util/send_email.py:20]: I think this error catching is dangerous. If something is going wrong we want to know, loud and clear, both on the diagnostics and the client side. You are also not really doing anything with the returned information later on, so I think it would be best just to remove the catch clause.

**Review by Drvanon — COMMENTED** [review]
I think I would really like for you to deal with the send_email comment before I merge it, but for the rest it's looking good.

**Review by Drvanon — COMMENTED** [review]
Looking good!

**Comment by Drvanon** [comment]
In my opinion this branch is now fully mergable.
